### PR TITLE
feat: add https proxy server implementation

### PR DESCRIPTION
**Pull request by bliuchak**
This PR represents simplification of https://github.com/apify/proxy-chain/pull/602. It contains changes related only to HTTPS proxy server support.

Note: I might miss some edge cases for HTTPS server proxy, however I think that current coverage is good enough and covers core logic fully (80/20) and is backward compatible (HTTP proxy server tests pass).

Closes: #603

### CODE REVIEW

File: src/chain.ts [hunk 1]
```diff
@@ -181,6 +193,8 @@ export const chain = (
 
         // Same here.
         sourceSocket.on('close', () => {
+            sourceSocket.unpipe(targetSocket);
+            targetSocket.unpipe(sourceSocket);
```
Comment (jirimoravcik) [inline src/chain.ts]: Can we comment on why is this needed? There are some more places with this as well.

File: src/chain.ts [hunk 2]
```diff
@@ -163,7 +167,13 @@ export const chain = (
...
         });
 
-        sourceSocket.write(isPlain ? '' : `HTTP/1.1 200 Connection Established\r\n\r\n`);
+        try {
+            sourceSocket.write(isPlain ? '' : `HTTP/1.1 200 Connection Established\r\n\r\n`);
+        } catch (error) {
+            targetSocket.destroy();
+            sourceSocket.destroy(error as Error);
+            return;
+        }
```
Comment (jirimoravcik) [inline src/chain.ts]: Please comment on why is try catch needed

File: src/server.ts [hunk 3]
```diff
@@ -96,10 +121,31 @@ export type PrepareRequestFunctionResult = {
...
+}
+
+export interface HttpServerOptions extends ServerOptionsBase {
+    serverType?: 'http';
+}
+
+export interface HttpsServerOptions extends ServerOptionsBase {
+    serverType: 'https';
+    httpsOptions: https.ServerOptions;
+}
```
Comment (jirimoravcik) [inline src/server.ts]: I believe it's preferred to use `type` instead of `interface`, can we do it here?

**Review by jirimoravcik — CHANGES_REQUESTED** [review]
Thanks for the PR, looks great, had a few comments

File: src/chain.ts (hunk 2 above)
Comment (bliuchak) [reply src/chain.ts]: Added.

File: README.md [hunk 4]
```diff
@@ -110,6 +110,99 @@ server.on('requestFailed', ({ request, error }) => {
...
 ```
 
+## Run a simple HTTPS proxy server
+
+This example demonstrates how to create an HTTPS proxy server with a self-signed certificate. The HTTPS proxy server works identically to the HTTP version but with TLS encryption.
+
+```javascript
+const fs = require('fs');
+const path = require('path');
+const ProxyChain = require('proxy-chain');
```
[the remark this answers is not in the document]
Comment (bliuchak) [reply README.md]: I was thinking about it too. It definitely makes sense. I'd migrate examples together with general migration to ESM in proxy-chain. At least right now our code is consistent (everywhere CJS is used) 😄

File: test/https-stress-test.js [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+            );
+        }
+
+        await Promise.all(promises);
+        await new Promise((r) => setTimeout(r, 500));
+
+        expect(allStats.length).to.equal(REQUESTS);
+
+        allStats.forEach((stats) => {
+            expect(stats).to.be.deep.equal({ srcTxBytes: 174, srcRxBytes: 93, trgTxBytes: 71, trgRxBytes: 174 })
```
[the remark this answers is not in the document]
Comment (bliuchak) [reply test/https-stress-test.js]: No, TLS overhead is not represented in these bytes (it's out of scope of this PR). These bytes represent only app-data.

The difference between `srcRxBytes` (93) and `trgTxBytes` (71) is due to hop-by-hop headers (like Proxy-Connection) being stripped when forwarding the request.

File: src/server.ts [hunk 6]
```diff
@@ -41,10 +42,34 @@ export const SOCKS_PROTOCOLS = ['socks:', 'socks4:', 'socks4a:', 'socks5:', 'soc
 const DEFAULT_AUTH_REALM = 'ProxyChain';
 const DEFAULT_PROXY_SERVER_PORT = 8000;
 
+const HTTPS_DEFAULTS = {
```
Comment (tobice) [inline src/server.ts]:
```suggestion
const HTTPS_DEFAULT_OPTIONS = {
```

(opt) Just more consistent naming across the code...

File: test/server.js [hunk 7]
```diff
@@ -1707,3 +1792,353 @@ describe('Socket error handler regression test', () => {
...
+    let server;
+
+    afterEach(async () => {
+        if (server) {
+            await server.close(true);
+            server = null;
+        }
+    });
+
+    it('handles TLS handshake failures gracefully and continues accepting connections', function (done) {
```
Comment (tobice) [inline test/server.js]: (opt) The `describe` has only one `it`. Maybe you can inline it?

File: test/server.js
[the remark this answers is not in the document]
Comment (tobice) [reply test/server.js]: Yep, you already have `https-stress-test.js`, so maybe the new nice https scenarios could  be stand-alone files as well?

File: test/https-stress-test.js [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+const requestPromised = (opts) => {
+    return new Promise((resolve, reject) => {
+        request(opts, (error, response, _) => {
+            if (error) {
+                return reject(error);
+            }
+            resolve(response);
+        });
+    });
+};
```
[the remark this answers is not in the document]
Comment (bliuchak) [reply test/https-stress-test.js]: Fixed.

File: test/https-stress-test.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (bliuchak) [reply test/https-stress-test.js]: Comment added.

File: src/server.ts [hunk 9]
```diff
@@ -174,11 +219,43 @@ export class Server extends EventEmitter {
...
+        // Keep legacy behavior (http) as default behavior.
+        this.serverType = options.serverType === 'https' ? 'https' : 'http';
+
+        if (options.serverType === 'https') {
+            if (!options.httpsOptions) {
+                throw new Error('httpsOptions is required when serverType is "https"');
+            }
+
+            // Apply secure TLS defaults (user options can override).
+            const secureDefaults: https.ServerOptions = {
```
[the remark this answers is not in the document]
Comment (bliuchak) [reply src/server.ts]: Fixed.

File: test/server.js
[the remark this answers is not in the document]
Comment (bliuchak) [reply test/server.js]: I've moved HTTPS specific tests into separate file. Any major refactoring I'd plan outside of this issue.

File: src/server.ts [hunk 10]
```diff
@@ -41,10 +42,34 @@ export const SOCKS_PROTOCOLS = ['socks:', 'socks4:', 'socks4a:', 'socks5:', 'soc
 const DEFAULT_AUTH_REALM = 'ProxyChain';
 const DEFAULT_PROXY_SERVER_PORT = 8000;
 
+const HTTPS_DEFAULT_OPTIONS = {
+    minVersion: 'TLSv1.2', // Disable TLS 1.0 and 1.1 (deprecated, insecure)
+    maxVersion: 'TLSv1.3', // Enable modern TLS 1.3
```
Comment (danpoletaev) [inline src/server.ts]: According to gemini here: _It is better to rely on Node.js defaults (controlled by minVersion: 'TLSv1.2'), which are
automatically updated for security, rather than maintaining a hardcoded list that might become obsolete._

Meaning we can remove `ciphers` and `maxVersion` 🤔

**Review by danpoletaev — APPROVED** [review]
We went through this in person! Looks good for me 👍
